Add GPP support for AMP endpoint (consent_type=4) (#3577)

postindustria-code opened the pull request:
[Issue #3577](https://github.com/prebid/prebid-server/issues/3577) 
- Add ConsentGPP constant (value 4) for GPP consent type
- Add GppSid field to AMP Params struct for gpp_sid parameter
- Create gpp.ConsentWriter in privacy/gpp package
- Update ReadPolicy to handle ConsentGPP case
- Add validateGppSid function for GPP SID validation
- Set regs.gpp and regs.gppsid in OpenRTB request

When consent_type=4:
- consent_string is copied to ORTB regs.gpp
- gpp_sid is validated and copied to regs.gppsid
- Invalid gpp_sid results in warning and nil gppsid

postindustria-code committed:
Add GPP support for AMP endpoint (consent_type=4) (#3577)

- Add ConsentGPP constant (value 4) for GPP consent type
- Add GppSid field to AMP Params struct for gpp_sid parameter
- Create gpp.ConsentWriter in privacy/gpp package
- Update ReadPolicy to handle ConsentGPP case
- Add validateGppSid function for GPP SID validation
- Set regs.gpp and regs.gppsid in OpenRTB request

When consent_type=4:
- consent_string is copied to ORTB regs.gpp
- gpp_sid is validated and copied to regs.gppsid
- Invalid gpp_sid results in warning and nil gppsid

postindustria-code committed:
Merge branch 'master' into amp-gpp-support

postindustria-code committed:
fix: package version

antoxas1986 reviewed (changes requested):
## Review: GPP support for AMP endpoint (consent_type=4)

Thanks for tackling #3577 — the feature is well-scoped and the unit tests pass. However, there's a correctness bug where a valid GPP consent string is silently dropped, plus a behavioral inconsistency with the rest of the codebase. Requesting changes; details and suggested fixes below.

---

### 🔴 Critical

**C1 — A valid `gpp` consent string is silently discarded when `gpp_sid` is malformed.**

The new `ConsentGPP` case in `amp/parse.go` returns *both* a warning and a valid writer when `gpp_sid` is invalid. But `overrideWithParams` in `endpoints/openrtb2/amp_auction.go` returns early on any non-nil warning, **before** calling `Write`:

```go
policyWriter, policyWriterErr := amp.ReadPolicy(ampParams, deps.cfg.GDPR.Enabled)
if policyWriterErr != nil {
    return []error{policyWriterErr}   // ← returns before policyWriter.Write(req)
}
if err := policyWriter.Write(req); err != nil { ... }
```

Because the warning is non-fatal (`errortypes.Warning`), the request still proceeds — but `Write` never ran, so `req.Regs.GPP` is never set. A publisher sending a **valid `gpp` string + a malformed `gpp_sid`** loses the privacy signal entirely, which is the opposite of the intended fail-open behavior. The "set gpp, leave gppsid nil" handling inside `ConsentWriter.Write` is unreachable from the AMP path in exactly the case it was written for.

This early-return is shared by the CCPA/TCF2 cases, but there the warning is about the *primary* consent string being invalid, so dropping it is defensible. For GPP the dropped value is *valid* and orthogonal to the bad field — hence critical. Note that `TestReadPolicyGPP` and `TestGppConsentWriterWrite` pass only because they exercise `ReadPolicy`/`Write` in isolation and never go through `overrideWithParams`.

**Suggested fix** — call `Write` first, then fold the warning into the accumulated `errors` slice instead of short-circuiting (mirrors how `setTargeting`/`setTrace` warnings are already collected lower in the function):

```go
policyWriter, policyWriterErr := amp.ReadPolicy(ampParams, deps.cfg.GDPR.Enabled)
if err := policyWriter.Write(req); err != nil {
    return []error{err}
}

var errors []error
if policyWriterErr != nil {
    errors = append(errors, policyWriterErr) // non-fatal warning; continue
}

// ... existing Timeout handling ...
if warn := setTargeting(req, ampParams.Targeting); warn != nil {
    errors = append(errors, warn)
}
// ... etc.
```

Please also add an `overrideWithParams`/handler-level test asserting that a valid `gpp` string + invalid `gpp_sid` results in `req.Regs.GPP` being set **and** a warning being returned.

---

### 🟠 Major

**M1 — Parse `gpp_sid` once instead of twice (also cleans up C1).**
`validateGppSid` calls `stringutil.StrToInt8Slice` purely to detect an error and discards the result; `ConsentWriter.Write` then parses the identical string again. Beyond the redundant work, the parse logic is split across two packages and can drift. Prefer parsing once in `ReadPolicy` and passing a typed slice to the writer — this also deletes `validateGppSid` and removes the swallowed error in m3:

```go
// amp/parse.go — ConsentGPP case
case ConsentGPP:
    gppSID, err := stringutil.StrToInt8Slice(ampParams.GppSid)
    if err != nil {
        warningMsg = fmt.Sprintf("GPP SID '%s' is not a valid comma-separated list of integers.", ampParams.GppSid)
    }
    rv = gpp.ConsentWriter{Consent: ampParams.Consent, GppSid: gppSID}
```

```go
// privacy/gpp/consentwriter.go
type ConsentWriter struct {
    Consent string
    GppSid  []int8
}

func (c ConsentWriter) Write(req *openrtb2.BidRequest) error {
    if req == nil {
        return nil
    }
    if req.Regs == nil {
        req.Regs = &openrtb2.Regs{}
    }
    if c.Consent != "" {
        req.Regs.GPP = c.Consent
    }
    if len(c.GppSid) > 0 {
        req.Regs.GPPSID = c.GppSid
    }
    return nil
}
```

**M2 — Inconsistent invalid-`gpp_sid` handling vs. `setuid.go`.**
`endpoints/setuid.go` treats a malformed `gpp_sid` as a fatal `400` (`"invalid gpp_sid encoding, must be a csv list of integers"`), while this PR makes it a non-fatal warning. A fail-open divergence is defensible for AMP, but it should be a deliberate, commented choice — and as written (combined with C1) it currently both warns *and* drops valid consent. Once C1 is fixed, please add a short comment in the `ConsentGPP` case noting AMP intentionally degrades gracefully (unlike setuid) so a future reader doesn't "correct" it.

---

### 🟡 Minor

**m3 — `Write` swallows its parse error.** `if err == nil { ... }` discards the error while `Write` declares an `error` return it never uses. Disappears entirely if you adopt M1's pre-parsed slice; otherwise either return it or document the intentional ignore.

**m4 — Test smell: branching on `test.name`.** In `privacy/gpp/consentwriter_test.go`, `if test.name != "nil request does not panic"` is brittle — renaming the case silently changes behavior. Use an explicit table field:

```go
tests := []struct {
    name   string
    nilReq bool
    // ...
}{ /* ... */ }

var req *openrtb2.BidRequest
if !test.nilReq {
    req = &openrtb2.BidRequest{}
}
```

**m5 — Duplicate test coverage.** `amp/parse_test.go::TestGppConsentWriterWrite` re-tests `gpp.ConsentWriter.Write`, which is already fully covered in `privacy/gpp/consentwriter_test.go`. Suggest removing it and instead extending `TestReadPolicyGPP` to assert the writer actually mutates the request (that would also have caught C1).

**m6 — `Consent` empty but `gpp_sid` present ⇒ nothing written.** `ReadPolicy` early-returns `NilPolicyWriter` before the switch when `Consent` is empty, so `consent_type=4&gpp_sid=2` (no consent string) writes neither `regs.gpp` nor `regs.gppsid`. Consistent with CCPA/GDPR and documented by the test — just please confirm this matches the #3577 spec, since a publisher might legitimately want to send only `gpp_sid`.

**m7 — Stale doc comment.** `ReadPolicy`'s doc comment still says "Returned policy writer could either be GDPR, CCPA or NilPolicy." — please add GPP. Also confirm the public AMP `consent_type` documentation is updated to list `4`.

---

### ✅ Positives
- `ConsentWriter.Write` correctly guards `req == nil` and lazily allocates `Regs`, matching the `ccpa`/`gdpr` writers.
- Reuses the shared `stringutil.StrToInt8Slice` (same helper `setuid.go` uses) rather than reinventing parsing.
- Good isolated table-driven coverage of `ParseParams` and `validateGppSid`.

Bottom line: the key blocker is C1 — please make the GPP string write independently of `gpp_sid` validity (ideally via M1's single-parse refactor), reconcile the warning-vs-fatal decision (M2), and add an endpoint-level test. `go build ./...` and the existing `amp` / `privacy/gpp` tests pass locally.

postindustria-code committed:
AMP GPP: fix dropped consent, single-parse gpp_sid, honor lone gpp_sid

- Apply the GPP consent writer even when gpp_sid is malformed, so a valid gpp
  string is no longer dropped (overrideWithParams previously returned on the
  warning before Write). CCPA/TCF2 drop-on-invalid behavior is preserved.
- Parse gpp_sid once in ReadPolicy and pass a typed []int8 to gpp.ConsentWriter;
  remove the redundant validateGppSid helper.
- Write regs.gppsid independently of consent_string per issue #3577 so a lone
  gpp_sid is honored (matches PBS-Java AMP behavior).
- Document AMP's fail-open handling vs setuid's 400; tidy up the tests.

postindustria-code commented:
Addressed the review feedback in `5d955aa6c`:

- **C1** — a valid gpp consent string is no longer dropped when gpp_sid is malformed. overrideWithParams now applies the GPP writer and collects the gpp_sid warning as non-fatal, instead of returning before Write. CCPA/TCF2 keep their existing drop-on-invalid behavior (still covered by `TestConsentWarnings`).
- **M1** — `gpp_sid` is parsed once in ReadPolicy; `gpp.ConsentWriter.GppSid` is now a typed `[]int8` and the redundant `validateGppSid` helper is removed.
- M2/m7 — documented AMP's intentional fail-open for `gpp_sid` (vs setuid.go's 400) and fixed the ReadPolicy doc comment.
- m4/m5 — test cleanups (removed the duplicate writer test; the nil request case uses an explicit field).
- m6 / issue #3577 — per §1.1/§1.2, `gpp_sid` is written to `regs.gppsid` independently of consent_string, so a lone `gpp_sid` is now honored (matches PBS-Java AmpRequestFactory). Added unit + endpoint tests.

One thing worth a maintainer decision — divergence from PBS-Java on a malformed gpp_sid + valid gpp: this PR keeps regs.gpp and only clears regs.gppsid (following #3577 §1.1/§1.2, which treat the two as independent). PBS-Java's AmpRequestFactory.createRegs instead drops both (gpp is gated on successful gpp_sid extraction). Happy to match Java if you prefer strict parity — let me know.

`go test ./amp/... ./privacy/gpp/... ./endpoints/openrtb2/...` passes; `build/vet/gofmt` clean.

postindustria-code committed:
Merge branch 'master' into amp-gpp-support

antoxas1986 reviewed (approved):
Thanks for the thorough fixes — I re-verified against the diff (not just the summary) and all the items check out:

- **C1**: fixed — `Write()` now runs unconditionally before the GPP warning is folded in, and CCPA/TCF2 still short-circuit as before.
- **M1**: fixed — `gpp_sid` parsed once, typed `[]int8` threaded through, `validateGppSid` removed.
- **M2/m7**: fixed — the fail-open-vs-setuid.go divergence is now documented, and the `ReadPolicy` doc comment covers GPP.
- **m3/m4/m6**: fixed as described.
- **m5**: mostly fixed — `TestReadPolicyGPP` still re-asserts some of the same `Write()`/`Regs` behavior covered in `consentwriter_test.go`, but that's cosmetic, not blocking.

Two small non-blocking nits for a follow-up (happy to take as a fast-follow or separate PR, not holding this up):
1. `gpp.ConsentWriter.Write()` unconditionally allocates `req.Regs = &openrtb2.Regs{}` even when there's nothing to write (e.g. malformed `gpp_sid` + empty consent). `ccpa`/`gdpr` writers only allocate inside their non-empty guards — worth matching that for consistency.
2. The `_, isGPP := policyWriter.(gpp.ConsentWriter)` type assertion in `overrideWithParams` works, but it's a bit fragile — if `ConsentWriter` is ever returned as a pointer in a future refactor, the assertion silently goes false and reintroduces the original C1 behavior with no compiler signal. Not asking for a change now, just flagging so it doesn't bite later.

Re: the PBS-Java parity question — let's keep the current independent-fields behavior (per #3577 §1.1/§1.2). No need to chase strict Java parity here.

LGTM, approving.
